### PR TITLE
Feat: add dipole consistency test

### PR DESCRIPTION
**Pull request by anyangml**
This PR is to add cross framework consistency test on DipoleFittingNet.

Known Limitations:

1. There are some mismatched keys in the serialized model, only common keys are tested.

### CODE REVIEW

File: deepmd/tf/fit/fitting.py [hunk 1]
```diff
@@ -231,7 +235,7 @@ def serialize_network(
                 # initialize the network if it is not initialized
                 fittings[network_idx] = FittingNet(
                     in_dim=in_dim,
-                    out_dim=1,
+                    out_dim=out_dim,
```
Comment (anyangml) [inline deepmd/tf/fit/fitting.py:238]: This is to ensure the serialize/deserialize is consistent when using `serialize_network`

File: source/tests/consistent/common.py [hunk 2]
```diff
@@ -252,9 +252,16 @@ def test_tf_consistent_with_ref(self):
         tf_obj = self.tf_class.deserialize(data1, suffix=self.unique_id)
         ret2, data2 = self.get_tf_ret_serialization_from_cls(tf_obj)
         ret2 = self.extract_ret(ret2, self.RefBackend.TF)
+        if tf_obj.__class__.__name__.startswith("Polar", "Dipole"):
+            # tf, pt serialization mismatch
+            common_keys = set(data1.keys()) & set(data2.keys())
+            data1 = {k: data1[k] for k in common_keys}
+            data2 = {k: data2[k] for k in common_keys}
```
Comment (anyangml) [inline source/tests/consistent/common.py:259]: There are some mismatching implementations between PT and TF:

-  Distinguish_type/mixed_type
- Statistics not yet implemented on PT
- sel_types/exclude_types

File: deepmd/dpmodel/fitting/invar_fitting.py [hunk 3]
```diff
@@ -163,6 +163,7 @@ def __init__(
     def serialize(self) -> dict:
         data = super().serialize()
         data["dim_out"] = self.dim_out
+        data["atom_ener"] = []
```
Comment (wanghan-iapcm) [inline deepmd/dpmodel/fitting/invar_fitting.py:166]:
```suggestion
        data["atom_ener"] = self.atom_ener
```

File: deepmd/dpmodel/fitting/invar_fitting.py [hunk 4]
```diff
@@ -139,6 +139,7 @@ def __init__(
             raise NotImplementedError("atom_ener is not implemented")
 
         self.dim_out = dim_out
+        self.atom_ener = (atom_ener,)
```
Comment (wanghan-iapcm) [inline deepmd/dpmodel/fitting/invar_fitting.py:142]: im confused by the assignment. it gives a `Tuple[List[float]]`?

File: deepmd/tf/fit/dipole.py
Comment (njzjz) [inline deepmd/tf/fit/dipole.py:1]: The training seems to be broken:

```sh
cd examples/water_tensor/dipole
dp train dipole_input.json
```

```
Traceback (most recent call last):
  File "/home/jz748/anaconda3/bin/dp", line 8, in <module>
    sys.exit(main())
  File "/home/jz748/codes/deepmd-kit/deepmd/main.py", line 787, in main
    deepmd_main(args)
  File "/home/jz748/codes/deepmd-kit/deepmd/tf/entrypoints/main.py", line 72, in main
    train_dp(**dict_args)
  File "/home/jz748/codes/deepmd-kit/deepmd/tf/entrypoints/train.py", line 162, in train
    _do_work(jdata, run_opt, is_compress)
  File "/home/jz748/codes/deepmd-kit/deepmd/tf/entrypoints/train.py", line 186, in _do_work
    model = DPTrainer(jdata, run_opt=run_opt, is_compress=is_compress)
  File "/home/jz748/codes/deepmd-kit/deepmd/tf/train/trainer.py", line 87, in __init__
    self._init_param(jdata)
  File "/home/jz748/codes/deepmd-kit/deepmd/tf/train/trainer.py", line 105, in _init_param
    self.model = Model(**model_param)
  File "/home/jz748/codes/deepmd-kit/deepmd/tf/model/tensor.py", line 247, in __init__
    TensorModel.__init__(self, "dipole", *args, **kwargs)
  File "/home/jz748/codes/deepmd-kit/deepmd/tf/model/tensor.py", line 59, in __init__
    super().__init__(
  File "/home/jz748/codes/deepmd-kit/deepmd/tf/model/model.py", line 690, in __init__
    self.fitting = Fitting(
  File "/home/jz748/codes/deepmd-kit/deepmd/utils/plugin.py", line 84, in __call__
    obj.__init__(*args, **kwargs)
TypeError: DipoleFittingSeA.__init__() missing 1 required positional argument: 'embedding_width'
```

while it has no error using `devel`.

Comment (njzjz) [reply deepmd/tf/fit/dipole.py:1]: This error is not caught by UT.

File: source/tests/tf/test_data_modifier_shuffle.py [hunk 5]
```diff
@@ -153,6 +153,7 @@ def _setUp_jdata(self):
                 },
                 "fitting_net": {
                     "type": "dipole",
+                    "embedding_width": 4,
```
Comment (njzjz) [inline source/tests/tf/test_data_modifier_shuffle.py:156]: The user input script should not be modified.

File: deepmd/tf/fit/dipole.py [hunk 6]
```diff
@@ -327,3 +334,61 @@ def get_loss(self, loss: dict, lr) -> Loss:
...
+            The serialized data
+        """
+        data = {
+            "var_name": "dipole",
+            "ntypes": self.ntypes,
+            "dim_descrpt": self.dim_descrpt,
+            "embedding_width": self.dim_rot_mat_1,
+            # very bad design: type embedding is not passed to the class
+            # TODO: refactor the class
+            "distinguish_types": True,
```
Comment (njzjz) [inline deepmd/tf/fit/dipole.py:353]: I though `distinguish_types` has been entirely removed by #3289. No? @wanghan-iapcm

**Review by njzjz — CHANGES_REQUESTED** [review]

File: deepmd/dpmodel/fitting/invar_fitting.py (hunk 4 above)
Comment (anyangml) [reply deepmd/dpmodel/fitting/invar_fitting.py:142]: typo, reverted.

File: source/tests/tf/test_data_modifier.py [hunk 7]
```diff
@@ -57,7 +57,9 @@ def _setUp(self):
             restart=None, init_model=None, log_path=None, log_level=30, mpi_log="master"
         )
         jdata = j_loader(INPUT)
-
+        jdata["model"]["fitting_net"]["embedding_width"] = jdata["model"]["descriptor"][
+            "neuron"
+        ][-1]
```
Comment (njzjz) [inline source/tests/tf/test_data_modifier.py:62]: This should not be required. `DPTrainer` should be able to add it automatically. If not, there is a bug.

File: source/tests/tf/test_data_modifier_shuffle.py [hunk 8]
```diff
@@ -59,7 +59,9 @@ def _setUp(self):
         )
         jdata = self._setUp_jdata()
         self._setUp_data()
-
+        jdata["model"]["fitting_net"]["embedding_width"] = jdata["model"]["descriptor"][
+            "neuron"
+        ][-1]
```
Comment (njzjz) [inline source/tests/tf/test_data_modifier_shuffle.py:64]: This should not be required.

File: source/tests/tf/test_data_modifier.py (hunk 7 above)
Comment (anyangml) [reply source/tests/tf/test_data_modifier.py:62]: right.

File: deepmd/tf/model/model.py [hunk 9]
```diff
@@ -693,6 +693,7 @@ def __init__(
                 spin=self.spin,
                 ntypes=self.descrpt.get_ntypes(),
                 dim_descrpt=self.descrpt.get_dim_out(),
+                embedding_width=self.descrpt.get_dim_rot_mat_1(),
```
Comment (njzjz) [inline deepmd/tf/model/model.py:696]: There is an issue here. `get_dim_rot_mat_1` is unavailable for `loc_frame`. The current behavior is to throw `NotImplementEderror`. The energy fitting doesn't need this argument, while `loc_frame` only supports the energy fitting.

I am not sure what the best design is here...

Comment (anyangml) [reply deepmd/tf/model/model.py:696]: how about we add an check here based on the fitting_net type?

Comment (njzjz) [reply deepmd/tf/model/model.py:696]: Btw, I think `se_e2_r` doesn't support it in theory.

**Review by njzjz — APPROVED** [review]

**Review by wanghan-iapcm — APPROVED** [review]

**Pull request merged**